### ISSUE TITLE
Containers not being killed when updated version is deployed

### ISSUE THREAD

**Issue by Fabio101** [nearform/nscale#74]
Hi Guys

I am noticing an issue which I am not sure is a result of my setup or Nscale itself but any insight would be appreciated:

* **When deploying an updated revision of a system it appears as if old containers are not being killed and thus new containers are either unable to start or starts a another running instance of the container (Albeit the updated version)**:
 * Containers that fail to start are usually due to *"Bind for 0.0.0.0:<PORT> failed: port is already allocated."*
 * Containers that do start run side by side with the older revision of the container like so:
```
CONTAINER ID        IMAGE                                                                                     COMMAND                CREATED             STATUS              PORTS                                      NAMES
a4a3af8f78a1        localhost:8011/microservices/app-cc28fdaf655240299569a468f9a42c39cb7eda8b:1426497642583   "/root/start_applica   15 seconds ago      Up 14 seconds       443/tcp, 80/tcp                            hopeful_hopper      
2b81f8009154        localhost:8011/microservices/app-a76a2a8e8882128e6f7821e6ba3cb3f8fe8873fd:1426494114016   "/root/start_applica   39 minutes ago      Up 39 minutes       443/tcp, 80/tcp                            sharp_wozniak
```

My Nscale Environment comprises of:
* CentOS 7
* Nodejs v0.10.33
* Docker version 1.5.0, build a8a31ef/1.5.0
* Nscale : 
 * nscale 		 0.14.0-pre.6
 * nscale-kernel 	 0.14.14
 * nscale-client 	 0.14.7
 * nscale-web 	 0.10.0
 * nscale-api 	 0.10.0

My Docker Container environments comprise of:
* CentOS 7
* Docker version 1.5.0, build a8a31ef/1.5.0

For what its worth, I had a test environment once upon a time which did unlink and un-deploy containers correctly and I did not experience the above problems, the version of nscale was:

 * nscale 		 0.14.0-pre.6
 * nscale-kernel 	 0.14.6
 * nscale-client 	 0.14.3
 * nscale-web 	 0.10.0
 * nscale-api 	 0.10.0

Could it be my setup? Could it be Nscale?

Thanks guys!

**Comment by mcollina** [comment]
I am using this tool everyday, and I am not seeing this.

Can you please run `nscale system analyze` from your system folder, and paste the output, alongside with the output of `docker ps`.

**Comment by Fabio101** [comment]
Thanks Mcollina

I noticed that the analyze displayed a really messed up version of my system... missing many container definitions etc...

Let me try fix this up as it is most likely the cause of the problem here.

I will revert with Feedback soon :)

Thanks

**Comment by McDonnellDean** [comment]
@Fabio101 Could you post your result (minus and sensitive info if applicable). I would like to see how the analyzer handled your version?

**Comment by Fabio101** [comment]
@McDonnellDean for what its still worth, please see the analyses below:

```
nscale@NSCALE ~/microservices $ nscale system analyze microservices production 
{
  "name": "microservices",
  "namespace": "microservices",
  "containerDefinitions": [
    {
      "name": "Machine",
      "type": "blank-container",
      "specific": {},
      "id": "85d99b2c-06d0-5485-9501-4d4ed429799c"
    },
    {
      "id": "web$2f9f7ddadc8bead84de4a74665085d362b1854ff",
      "name": "web$2f9f7ddadc8bead84de4a74665085d362b1854ff",
      "type": "docker",
      "specific": {
        "repositoryUrl": "",
        "buildScript": "",
        "arguments": "",
        "buildHead": 0,
        "dockerImageId": "6b74fbfb128d81ad4adc73f0835c52cd2a4e2a882f0bad0cd92f30ebaacadb66",
        "imageTags": [
          "localhost:8011/nscaledemo/web-2f9f7ddadc8bead84de4a74665085d362b1854ff:1426239126948",
          "localhost:8011/nscaledemo/web-2f9f7ddadc8bead84de4a74665085d362b1854ff:latest"
        ]
      }
    }
  ],
  "topology": {
    "containers": {
      "logging-ca9562fa": {
        "id": "logging-ca9562fa",
        "containerDefinitionId": "85d99b2c-06d0-5485-9501-4d4ed429799c",
        "containedBy": "logging-ca9562fa",
        "contains": [],
        "type": "blank-container",
        "specific": {
          "ipaddress": "localhost"
        }
      }
    },
    "name": "production"
  }
}
```

**HOWEVER!**

I have solved my problem...

Inside the /home/nscale/.nscale/config/.config.json the analysis block contained:
```
"analysis": {
      "require": "nscale-local-analyzer",
      "specific": {
      }
    }
  },
```

But what I actually required was...
```
"analysis": {
      "require": "nscale-direct-analyzer",
      "specific": {
        "user": "nscale",
        "identityFile": "/home/nscale/.ssh/id_rsa"
      }
    }
  },
```

Basically the local analyzer is defined by default during installation of Nscale which I was not aware of (My initial environment was setup by someone else and that person must have missed telling me that this was a specific configuration requirement for our environment)

Now my analyses work correctly as well as deployment and redeployment of updated revision (Containers are killed and replaced accordingly.)

Thanks guys!

**Comment by mcollina** [comment]
This will be solved in the next release 0.15, where we will have automatic
detect of the right analyzer to use.

Il giorno lun 16 mar 2015 alle ore 13:02 Fabio Pinto <
notifications@github.com> ha scritto:

**Comment by Fabio101** [comment]
Perfect, looking forward to it!
